### DISCUSSION TITLE
How to do things with ES Modules

### DISCUSSION THREAD

**Discussion by stuartcusackie** [tailwindlabs/tailwindcss#14224] [Help]
ES Module usage doesn't seem to be covered in the Tailwind docs. I have two main problems when switching to an ES Module based project:

### DEFAULT THEME:

I used to do this:
```
const defaultTheme = require('tailwindcss/defaultTheme')

fontFamily: {
  sans: ['Raleway', ...defaultTheme.fontFamily.sans],
  serif: [...defaultTheme.fontFamily.serif]
},
```

But with ES, which I'm kind of stuck with, you get:
```
require is not defined in ES module scope, you can use import instead 
```

So I tried this:
```
import defaultTheme from "tailwindcss";
```

but that results in:
```
Cannot read properties of undefined (reading 'sans')
```

### CUSTOM PLUGINS

I used to write plugins as described in the docs like this:

```
const plugin = require("tailwindcss/plugin")

module.exports = plugin(function ({ addUtilities }) {
	const utils = {
		'.w-edge': {
			position: 'relative',
			margin: '0 auto',
			width: '100vw',
			left: '50%',
			transform: 'translateX(-50%)'
		}
	}

  addUtilities(utils, ["responsive"])
})
```

but how do you convert this to an ES module using export default?

```
// This doesn't work obviously
exports default = plugin(function ({ addUtilities }) {
```

Thanks

**Comment by wongjn** [comment]
## Default Theme

Try:

```js
import defaultTheme from 'tailwindcss/defaultTheme';
```

## Custom plugins

Try:

```js
export default plugin(function ({ addUtilities }) {
```

**Comment by stuartcusackie** [comment]
Not quite but you pointed me in the right direction. The only thing missing was the .js extension, which seems to be required with the import statement. I suppose this depends on your compiler. I'm not using Vite, it's an unusual webpack bundler.

```
// tailwind.config.js
import defaultTheme from 'tailwindcss/defaultTheme.js';
import utilityWEdge from './resources/tailwind/utilities/w-edge.js';

plugins: [
  utilityWEdge
]
```

```
// resources/tailwind/utilities/w-edge.js
import plugin from 'tailwindcss/plugin.js';

export default plugin(function ({ addUtilities }) {
	const utils = {
		'.w-edge': {
			position: 'relative',
			margin: '0 auto',
			width: '100vw',
			left: '50%',
			transform: 'translateX(-50%)'
		}
	}

  addUtilities(utils, ["responsive"])
})
```

Thanks for the help!

**Reply by wongjn** [reply]
Seems to work fine in [this Stackblitz project](https://stackblitz.com/edit/vitejs-vite-gsvfs5?file=tailwind.config.js). Consider checking you have Tailwind installed as an NPM module. Otherwise, consider providing a project that reproduces the undesired behavior.

**Reply by stuartcusackie** [reply]
It's working! I think you misread my response.
Thanks

**Reply by wongjn** [reply]
I replied to your pre-edit post:

> Afraid not:
>
> ```js
> import defaultTheme from 'tailwindcss/defaultTheme';
> ```
> ```
> Cannot find module '[secure]/node_modules/tailwindcss/defaultTheme' imported from [secure]/tailwind.config.js
> ```

**Reply by stuartcusackie** [reply]
ahah, sorry. I see. All fixed now. Thanks again.
